src/chains/ethereum/ethereum/src/helpers/simulations.ts

```diff
@@ -0,0 +1,582 @@
+      stateDiff: { [slot: string]: string };
+    }>;
+
+export type CallOverrides = {
+  [address: string]: CallOverride;
+};
+
+export type CreateAccessListResult = {
+  accessList: AccessList;
+  gasUsed: Quantity;
+};
+
+/**
+ * Stripped down version of the type from EthereumJs
+ */
+interface VMRunCallOpts {
+  caller?: EthereumJsAddress;
+  data?: Buffer;
+  gasPrice?: BN;
+  gasLimit?: BN;
+  to?: EthereumJsAddress;
+  value?: BN;
+  block?: any;
+}
+
+type CloneVmOptions = {
+  blockchain: Blockchain;
+  common: Common;
+  simulationBlock: Block;
+};
+
+type CloneVmResult = {
+  vm: VM;
+  stateManager: DefaultStateManager;
+  stateTrie: GanacheTrie;
+};
+
+type SimulatorCommonOptions = CloneVmOptions & {
+  options: EthereumInternalOptions;
+  emitStepEvent: boolean;
+  transaction: SimulationTransaction;
+};
+
+type RunCallOptions = SimulatorCommonOptions & {
+  overrides?: CallOverrides;
+};
+
+type CreateAccessListOptions = SimulatorCommonOptions;
+
+type RunCallSetupOptions = {
+  common: Common;
+  stateTrie: GanacheTrie;
+  stateManager: DefaultStateManager;
+  transaction: SimulationTransaction;
+  overrides?: CallOverrides;
+};
+
+type RunCallSetupResult = {
+  runCallOpts: VMRunCallOpts;
+  accessListExclusions: EthereumJsAddress[];
+  addressesOnlyStorage: EthereumJsAddress[];
+  intrinsicGas: bigint;
+};
+
+type InternalRunCallOptions = {
+  blockchain: Blockchain;
+  runCallOpts: VMRunCallOpts;
+  vm: VM;
+  options: EthereumInternalOptions;
+  emitStepEvent: boolean;
+};
+
+const toLightEJSAddress = (address?: Address): EthereumJsAddress => {
+  if (address) {
+    const buf = address.toBuffer();
+    return { buf, equals: (a: { buf: Buffer }) => buf.equals(a.buf) } as any;
+  } else {
+    return null;
+  }
+};
+
+/**
+ * Calls the StateManager's `addWarmedAddress` function for the transaction
+ * caller, all precompiles, and the to address, if applicable.
+ * @param stateManager
+ * @param caller
+ * @param to
+ * @returns An array of precompile addresses
+ */
+const warmDefaults = (
+  stateManager: DefaultStateManager,
+  caller: EthereumJsAddress,
+  to?: EthereumJsAddress
+) => {
+  let precompiles: EthereumJsAddress[] = [];
+  // handle Berlin hardfork warm storage reads
+  precompiles = warmPrecompiles(stateManager);
+  stateManager.addWarmedAddress(caller.buf);
+  if (to) stateManager.addWarmedAddress(to.buf);
+
+  return precompiles;
+};
+
+const warmAccessList = (
+  stateManager: DefaultStateManager,
+  accessList: AccessList
+) => {
+  for (const { address, storageKeys } of accessList) {
+    const addressBuf = Address.toBuffer(address);
+    stateManager.addWarmedAddress(addressBuf);
+    for (const slot of storageKeys) {
+      stateManager.addWarmedStorage(addressBuf, Data.toBuffer(slot, 32));
+    }
+  }
+};
+
+const validateStorageOverride = (
+  slot: string,
+  value: string,
+  fieldName: string
+) => {
+  // assume Quantity will handle other types, these are just special string cases
```

Comment:
Looks like you actually use Data, not Quantity.
```suggestion
  // assume Data will handle other types, these are just special string cases
```

Does it make sense to even do this validation? As you say, it's being done within `Data.from()` anyways - shall we just validate the length?

Reply:
I believe what I was saying (which is hard to gather from the context, so I'll fix) is that the _other_ overrides are Quantities, which have their own validation. However, the storage slot/values are "string cases" (aka Data), and require their own separate validation. The point of this specific class is to have the same errors that Geth does, which might be overkill.

Regardless, this override stuff was a PR from before so I don't want to make too many changes without an issue and a separate PR and properly thinking it through. I've just moved the existing code here.